invenio_rdm_records/records/dumpers.py

```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020 CERN.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""DataCite-based data model for Invenio."""
+
+import calendar
+from datetime import date
+
+from edtf import parse_edtf
+from edtf.parser.edtf_exceptions import EDTFParseException
+from invenio_records.dumpers import ElasticsearchDumperExt
+
+
+class EDTFDumperExt(ElasticsearchDumperExt):
+    """Elasticsearch dumper extension for EDTF dates support.
+
+    It adds two fields: publication_date_start and publication_date_end,
+    which correspond to the lower strict and upper strict bounds. This
+    are required for sorting. In an ascending sort (most recent last),
+    sort by publication_date_start to get a natural sort order. In a
+    descending sort (most recent first), sort by publication_date_end.
+    """
+
+    def __init__(self, field):
+        """Constructor.
+
+        :param field: dot separated path to the EDTF field to process.
+        """
+        super(EDTFDumperExt, self).__init__()
+        self.field_path = field.split('.')
+
+    def _dump(self, path, data):
+        """Find recursively and dump EDTF field."""
+        key = path[0]
```

Comment:
Better use https://github.com/inveniosoftware/invenio-records/blob/master/invenio_records/dictutils.py#L70 (as then we have consistent lookup strategy for field paths.

```python
keys = parse_lookup_key(self.field_path)
data = dict_lookup(data, keys, parent=True)
key = keys[-1]
pd = parse_edtf(data[key])
data[f'{key}_start'] = ...
```

Reply:
@lnielsen this would work for lookup and parse of the path. however, then we would set the created field in the root of the document. e.g.

```json
{ "metadata: {
   "publication_date": '2021'
  }
}
```

Would become

```json
{ "metadata: {
   "publication_date": "2021"
  },
 "publication_date_start": "2021-01-01",
 "publication_date_end": "2021-12-31",
}
```

Instead of:

```json
{ "metadata: {
   "publication_date": "2021",
   "publication_date_start": "2021-01-01",
   "publication_date_end": "2021-12-31",
  }
}
```

Taking into account that we have at the moment a search problem i.e. the search() method makes ES search in *every* field, including e.g. `access`, and not just inside `metadata`; once we fix that problem, this would not allow search on this type of fields (aggregation, facet and sorting would still be possible).

If search on this fields is not desired this approach would work.

Otherwise:
a- Create a `dict_setup` method (would mean looping twice on the array/field path)
b- Keep this, but inconsistent
c- Assume no search over that content

Reply:
Implemented https://github.com/inveniosoftware/invenio-rdm-records/pull/194/files#diff-3062ab2ea652c3ad7e17249676a44495cdc1eaf6f3c7b661c071c94dfccf06aaR41

Reply:
See a solution in https://github.com/inveniosoftware/invenio-rdm-records/pull/201